### ISSUE TITLE
[MPS] Incorrect output from convolution ops with large dimensions

### ISSUE THREAD

**Issue by hvaara** [pytorch/pytorch#142836]
### 🐛 Describe the bug

#140726 attempted to reintroduce support for convolutions with `out_channels` > 2**16. While this appeared to work for `Conv1d`, it introduced a regression in `Conv2d` (see https://github.com/pytorch/pytorch/issues/142515#issuecomment-2533495363). It remains unclear whether `Conv3d` is affected.

The issue results in a silent correctness bug. A guard was added in #129484 to prevent this, and we should reinstate the guard for the affected operations.

### TODO

- [ ] Investigate which ops are currently broken.
- [ ] Add additional tests with large dimensions for all operations on the affected code path.
- [ ] Reinstate the guard for broken operations on all macOS versions.
- [ ] Ensure the fix is included in the v2.6.0 release.

Sorry for the breakage!

### Versions

PyTorch version: 2.6.0a0+gite1196df
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.1.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: version 3.31.1
Libc version: N/A

Python version: 3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:19:53) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] mypy==1.13.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.2
[pip3] onnx==1.17.0
[pip3] onnxscript==0.1.0.dev20240817
[pip3] optree==0.13.0
[pip3] torch==2.6.0a0+gite1196df
[pip3] torchaudio==2.5.0a0+ba696ea
[pip3] torchvision==0.20.0a0+e9a3213
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.6.0a0+gite1196df           dev_0    <develop>
[conda] torchaudio                2.5.0a0+ba696ea           dev_0    <develop>
[conda] torchvision               0.20.0a0+e9a3213           dev_0    <develop>

**Comment by hvaara** [comment]
I volunteer as tribute -- feel free to assign me.

@pytorchbot label "module: regression" "module: correctness (silent)" "module: mps" "high priority"

**Comment by skotapati** [comment]
Hi @hvaara thanks for filing this issue, I'm able to reproduce the error with your script. I'll mark you as the assignee for now, but please let us know if you're no longer available and I'll take a look myself

**Comment by hvaara** [comment]
@skotapati You're absolutely welcome to pitch in if you're interested! There are two somewhat distinct tasks to address here:

1. **Short-term:** Reinstate the guard to prevent silent correctness errors from regressing (the one I previously removed). Some operations/parameter combinations seem unaffected, so our first step is identifying the failure modes. This involves adding `xfail` unit tests and defining an appropriate predicate in https://github.com/pytorch/pytorch/blob/d9d7cca009ba9a79f3662a3de057a081163b95f6/aten/src/ATen/native/mps/operations/Convolution.mm#L168-L173. For example, `nn.Conv1d` appears to be unaffected. Given the high visibility of errors from this operation - especially in popular TTS libraries with large channel counts - it’s important not to overapply the guard unnecessarily.  
   
2. **Long-term:** Address the underlying failure modes. As mentioned by @jhavukainen in https://github.com/pytorch/pytorch/pull/129484#issue-2373168517, this has been identified as future work and requires changes on the kernel side.

I’ve already started on item 1, aiming for a reviewable version by the end of the week. Feel free to reassign if needed, and if I haven’t made progress, don’t hesitate to give me a nudge.

I haven’t started on item 2 yet. I suggest opening a separate issue to track the long-term fix. @jhavukainen, was the plan to resolve this within MPS proper or within PyTorch? Addressing it in MPS seems like a cleaner solution, as it would reduce the need for downstream projects (ie. dependents on MPS; eg. PyTorch) to duplicate the implementation.

**Comment by jhavukainen** [comment]
@hvaara On point 2: Yes the long term fix will appear in a new version of an MPS release eventually once it ships with a newer OS version. Once that happens we can limit the raising the guard to only OS versions before.

I can also confirm that the fix landed already in MacOS 15.1 should have sorted out the problem for Conv1D. My intention was to get a fix fir all the convolution dimension variants but evidently that did not come to pass. So I've filed new internal issues to address the remaining problems with larger convolutions in 2D (and I presume 3D).

**Comment by kulinseth** [comment]
@hvaara , we will come back with if there is a solution at Framework level we can support > 2**16 out_channels
